base/compiler/tfuncs.jl

```diff
@@ -1704,6 +1714,85 @@ function _builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1}, @nospecializ
     return false
 end
 
+# known to be always effect-free (in particular nothrow)
+const _PURE_BUILTINS = Any[tuple, svec, ===, typeof, nfields]
+
+# known to be effect-free (but not necessarily nothrow)
+const _EFFECT_FREE_BUILTINS = [
+    fieldtype, apply_type, isa, UnionAll,
+    getfield, arrayref, const_arrayref, isdefined, Core.sizeof,
+    Core.kwfunc, Core.ifelse, Core._typevar, (<:),
+    typeassert, throw, arraysize
+]
+
+const _CONSISTENT_BUILTINS = Any[
+    tuple, # tuple is immutable, thus tuples of egal arguments are egal
+    ===,
+    typeof,
+    nfields,
+    fieldtype,
+    apply_type,
+    isa,
+    UnionAll,
+    Core.sizeof,
+    Core.kwfunc,
+    Core.ifelse,
+    (<:),
+    typeassert,
+    throw
+]
+
+const _SPECIAL_BUILTINS = Any[
+    Core._apply_iterate
+]
+
+function builtin_effects(f::Builtin, argtypes::Vector{Any}, rt)
+    if isa(f, IntrinsicFunction)
+        return intrinsic_effects(f, argtypes)
+    end
+
+    @assert !contains_is(_SPECIAL_BUILTINS, f)
+
+    nothrow = false
+    if (f === Core.getfield || f === Core.isdefined) && length(argtypes) >= 3
+        # consistent if the argtype is immutable
+        if isvarargtype(argtypes[2])
+            return Effects(Effects(), effect_free=ALWAYS_TRUE, terminates=ALWAYS_TRUE)
+        end
+        s = widenconst(argtypes[2])
+        if isType(s) || !isa(s, DataType) || isabstracttype(s)
+            return Effects(Effects(), effect_free=ALWAYS_TRUE, terminates=ALWAYS_TRUE)
```

Comment:
@Keno Isn't the `effect_free` here wrong if the first argument could be a module?

```julia
julia> getfield(Main, :+)
+ (generic function with 206 methods)

julia> a + b = 1
ERROR: error in method definition: function Base.+ must be explicitly imported to be extended
Stacktrace:
 [1] top-level scope
   @ none:0
 [2] top-level scope
   @ REPL[2]:1
```

Reply:
we don't consider binding resolution to be a formal effect, currently anyways

Reply:
Ah, wait, that would mean we couldn't constant fold any bindings owned by other modules if they haven't been accessed before.